rubicon/objc/objc.py

```diff
@@ -1644,3 +1660,201 @@ def finalize(self, cmd) -> None:
             anObject = get_instance_variable(self, 'observed_object', objc_id)
             ObjCInstance._cached_objects.pop(anObject, None)
             send_super(self, 'finalize')
+
+
+class ObjCBlockStruct(Structure):
+    _fields_ = [
+        ('isa', c_void_p),
+        ('flags', c_int),
+        ('reserved', c_int),
+        ('invoke', CFUNCTYPE(c_void_p, c_void_p)),
+        ('descriptor', c_void_p),
+    ]
+
+
+def create_block_descriptor_struct(has_helpers, has_signature):
+    descriptor_fields = [
+        ('reserved', c_ulong),
+        ('size', c_ulong),
+    ]
+    if has_helpers:
+        descriptor_fields.extend([
+            ('copy_helper', CFUNCTYPE(c_void_p, c_void_p, c_void_p)),
+            ('dispose_helper', CFUNCTYPE(c_void_p, c_void_p)),
+        ])
+    if has_signature:
+        descriptor_fields.extend([
+            ('signature', c_char_p),
+        ])
+    return type(
+        'ObjCBlockDescriptor',
+        (Structure, ),
+        {'_fields_': descriptor_fields}
+    )
+
+
+def cast_block_descriptor(block):
+    descriptor_struct = create_block_descriptor_struct(block.has_helpers, block.has_signature)
+    return cast(block.struct.contents.descriptor, POINTER(descriptor_struct))
+
+
+AUTO = object()
+_SIGNATURE_CACHE = {}
+
+
+def _strip_nums(siglist):
+    while siglist and siglist[0] in '0123456789':
+        siglist.popleft()
+
+
+SIMPLE_SIGNATURE_TYPES = {
+    'c': c_char,
+    'i': c_int,
+    's': c_short,
+    'l': c_long,
+    'q': c_longlong,
+    'C': c_ubyte,
+    'I': c_uint,
+    'S': c_ushort,
+    'L': c_ulong,
+    'Q': c_ulonglong,
+    'f': c_float,
+    'd': c_double,
+    'B': c_bool,
+    'v': c_void_p,
+    '*': c_char_p,
+    '#': Class,
+    ':': SEL,
+}
```

Comment:
#53 adds a `ctypes_for_method_encoding` function to decode a method signature into its return/argument types. This should also work for block signatures, which are in exactly the same format as method signatures (as far as I can tell).

Previously we used some Objective-C runtime functions to get the return/argument types from method signatures, but they only work with a `Method` object, and not with a "standalone" signature like we have for blocks. So I wrote a similar function in Python, based on the existing type decoding infrastructure, which works with a simple bytestring instead of a `Method` object.